Add viewMode toggle for condensed resource table

pascalwengerter opened the pull request:
## Description
Adds route-query based toggle between "normal" and "condensed" resource table

## Related Issue
- Fixes part of #6380

## Screenshots (if appropriate):
<img width="810" alt="Screenshot 2022-11-15 at 22 34 49" src="https://user-images.githubusercontent.com/16822008/202029210-13230f96-4a0c-408e-bf1f-f8dc8a371b54.png">

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Open tasks:
- [ ] Review/feedback/approval by @tbsbdr / @kulmann / @elizavetaRa

kulmann commented:
I have merged https://github.com/owncloud/web/pull/7940 today - could you rebase this PR and include all your design system changes directly in web? 😁

pascalwengerter commented:
> ~I have merged #7940 today - could you rebase this PR and include all your design system changes directly in web?~ nevermind, wanted to post this in #7991

Can't hurt though 😉

kulmann commented:
Just checked it out, my impression is that the font size is already too small. But it might just be that I'm not "scientific enough" for the scientific mode. 😁

pascalwengerter commented:
> Just checked it out, my impression is that the font size is already too small. But it might just be that I'm not "scientific enough" for the scientific mode. 😁

What does the scientific corner say? @elizavetaRa 🤓

tbsbdr commented:
## decrease whitespace
Let's only decrease whitespace, to achieve the goal of showing more lines in the viewport.

Example how this looks as dropbox:

<img width="1511" alt="screenshot_000011" src="https://user-images.githubusercontent.com/26610733/203536766-d5facf70-0f0a-432e-a59a-aa79acaafcd5.png">
<img width="1510" alt="screenshot_000010" src="https://user-images.githubusercontent.com/26610733/203536773-dd2a8160-4059-43d0-8bb4-66c63f287115.png">

let's **not** decrease fonstsize, because:
- this should be done ex. via browser (ctl+ "-") or in the future in the ocis user settings
- only decreasing fontsize in the fileslist and not in the other ui elements messes up the overal design impression

tbsbdr commented:
@pascalwengerter created a custom menu-line-condensed.svg
 
`menu-line-condensed.svg`
```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" width="24" height="24">
<path d="M3 5H21V7H3V5ZM3 9H21V11H3V9ZM3 17H21V19H3V17Z" fill="black"/>
<path d="M21 13H3V15H21V13Z" fill="black"/>
</svg>
```

Preview:
<img width="22" alt="screenshot_000012" src="https://user-images.githubusercontent.com/26610733/203560503-aca4dbcc-f866-4446-91a6-ecb4ecb6b8d3.png">

pascalwengerter commented:
@tbsbdr @kulmann please re-review

pascalwengerter commented on packages/design-system/changelog/unreleased/enhancement-condensed-menu-item:
```diff
@@ -0,0 +1,5 @@
+Enhancement: Add condensed menu icon 
```
@kulmann not sure if the changelog in the design system still gets updated individually or if there's only one in the repo root from now on?

dschmidt replied:
We need to maintain a separate changelog, because we want to be able to release independently with non aligned version numbers

pascalwengerter replied:
Good, that's what I assumed (and hence the changelog entry^^)

kulmann replied:
yes, this ^

elizavetaRa commented:
Nice feature, indeed asked by our users. Looking forward to it!

dschmidt commented on packages/web-app-files/src/composables/viewMode/constants.ts:
```diff
@@ -0,0 +1,4 @@
+export abstract class ViewModeConstants {
+  static readonly viewModeDefault: string = 'resource-table'
+  static readonly viewModeQueryName: string = 'view-mode'
+}
```
Not a huge fan of this use class as enum pattern, I'd rather use typescript types for this nowadays

pascalwengerter replied:
I've followed what I found in another composable basically, idk if refactoring my/the other solution independently is within the scope of this PR?

kulmann replied:
I agree with both of you 😆 Would like to get rid of it, but also fine in a separate PR.

kulmann commented on packages/web-app-files/src/components/FilesList/ResourceTable.vue:
```diff
@@ -1,6 +1,9 @@
-    :class="hoverableQuickActions && 'hoverable-quick-actions'"
+    :class="[
+      hoverableQuickActions && 'hoverable-quick-actions',
+      { condensed: viewMode === 'resource-table-condensed' }
```
Use the `const` please

kulmann commented on packages/web-app-files/src/components/FilesList/ResourceTable.vue:
```diff
@@ -318,6 +321,10 @@ export default defineComponent({
     },
+    viewMode: {
+      type: String,
+      default: 'resource-table'
```
Use the `const` please

kulmann commented on packages/web-app-files/src/composables/viewMode/useViewMode.ts:
```diff
@@ -0,0 +1,15 @@
+    name: ViewModeConstants.viewModeQueryName,
+    defaultValue: ViewModeConstants.viewModeDefault
+  })
+  return computed(() => String(unref(perPageQuery)))
```
```suggestion
  return computed(() => queryItemAsString(unref(perPageQuery)))
```
+ import from `web-pkg/src/composables`

kulmann reviewed (changes requested):
UX feels good! 🥳 

Could you please add unit tests for the view mode selection, as implemented in `ViewOptions.vue`? Added some comments.

pascalwengerter commented on packages/web-app-files/src/components/AppBar/ViewOptions.vue:
```diff
@@ -1,5 +1,19 @@
+        :appearance="viewModeCurrent === 'resource-table-condensed' ? 'filled' : 'outline'"
+        @click="setViewMode('resource-table-condensed')"
+      >
+        <oc-icon name="menu-line-condensed" fill-type="none" />
```
Done, good idea!

pascalwengerter commented on packages/web-app-files/src/composables/viewMode/useViewMode.ts:
```diff
@@ -0,0 +1,15 @@
+    return computed(() => unref(options))
+  }
+
+  const perPageQuery = useRouteQueryPersisted({
```
Done, thx 🤦🏽

pascalwengerter commented on packages/web-app-files/src/components/AppBar/ViewOptions.vue:
```diff
@@ -1,5 +1,19 @@
+  <div class="oc-flex oc-flex-middle">
+    <div class="oc-button-group oc-visible@s oc-mr-s">
+      <oc-button
+        :appearance="viewModeCurrent === 'resource-table-condensed' ? 'filled' : 'outline'"
```
Done, good point. Took me way longer than necessary because I was wondering what `const` you were talking about exactly^^. Please take a look at https://github.com/owncloud/web/pull/7976/commits/e14e4645eba41592c27f19e10463cc465e2ef3e8 and let me know which of the two constant variants you prefer! (as in `ViewModeConstants.viewModeDefault` vis `ViewModeConstants.default`, e.g.)

pascalwengerter commented:
> UX feels good! 🥳
> 
> Could you please add unit tests for the view mode selection, as implemented in `ViewOptions.vue`? Added some comments.

Thanks for the review, I've addressed all points AFAIK and added a unit test for the button group, basically checking whether the correct button gets shown as outlined/filled after clicking the respective option - is that sufficient for now?

pascalwengerter commented:
@kulmann could I get a re-review? :)

kulmann commented:
> @kulmann could I get a re-review? :)

Yes, but I'm still waiting for internal UX feedback from @tbsbdr 🙈

tbsbdr commented:
awesome! (was blocked as I did not manage to start it locally. Benedikt helped me)
I would have used the menu-line icon for regular-view mode and the [condensed](https://github.com/owncloud/web/pull/7976#issuecomment-1325083772) icon for condensed-view:

## Would love to see it like this
![screenshot_000053](https://user-images.githubusercontent.com/26610733/206486780-60ab4978-3968-4fb9-9320-c1213051f210.png)

## menu-line icon for regular-view mode
![screenshot_000054](https://user-images.githubusercontent.com/26610733/206488061-89530c9a-7805-4806-a9aa-7bdda3ec577d.png)

pascalwengerter commented:
> awesome! (was blocked as I did not manage to start it locally. Benedikt helped me) I would have used the menu-line icon for regular-view mode and the [condensed](https://github.com/owncloud/web/pull/7976#issuecomment-1325083772) icon for condensed-view:
> ## Would love to see it like this
> 
> ![screenshot_000053](https://user-images.githubusercontent.com/26610733/206486780-60ab4978-3968-4fb9-9320-c1213051f210.png)
> ## menu-line icon for regular-view mode
> 
> ![screenshot_000054](https://user-images.githubusercontent.com/26610733/206488061-89530c9a-7805-4806-a9aa-7bdda3ec577d.png)

Done!

kulmann commented:
@tbsbdr found a kind of UX flaw: the views `Favorites` (oc10 only at the moment), `Deleted files`, `Shared with others`, `Shared via link` and `Search results` have a line below the folder name for displaying the parent folder name/link. For those views the condensed and regular list view mode are identical. I'm torn between (a) not showing the condensed view mode button at all on those pages, or (b) showing the view mode button, knowing that it will not have any effect in the default theme. What do you think @pascalwengerter ?

kulmann commented on packages/web-app-files/tests/unit/components/AppBar/ViewOptions.spec.js:
```diff
@@ -95,4 +97,46 @@ describe('ViewOptions', () => {
   })
+
+  it('initially shows normal resource-table by default', () => {
+    const wrapper = shallowMount(ViewOptions, {
```
for future tests (not blocking this PR): we introduced a new way of handling mocks in unit tests. See for example https://github.com/owncloud/web/blob/5667b5fd9b58f7533d5866b27524afe3d344a59e/packages/web-app-files/tests/unit/views/spaces/DriveResolver.spec.ts#L56

The functions creating the wrappers now return an object, containing mocks, storeOptions and the wrapper. That gives you the chance to manipulate mocks within the test code.

kulmann reviewed (approved):
UX feels good, code is also good now. One hint about tests (ok for later, see comment). My feeling is that we can handle my previous comment in a followup. Let's merge this as first iteration of the condensed view mode.
